donkeycar/parts/oled.py

```diff
@@ -1,145 +1,430 @@
+        try:
+            if OLEDPart.get_network_interface_state(interface) == 'down':
+                return None
+            cmd = "ifconfig %s | grep -Eo 'inet (addr:)?([0-9]*\.){3}[0-9]*' | grep -Eo '([0-9]*\.){3}[0-9]*' | grep -v '127.0.0.1'" % interface
+            return subprocess.check_output(cmd, shell=True).decode('ascii')[:-1]
+        except:
             return None
-        cmd = "ifconfig %s | grep -Eo 'inet (addr:)?([0-9]*\.){3}[0-9]*' | grep -Eo '([0-9]*\.){3}[0-9]*' | grep -v '127.0.0.1'" % interface
-        return subprocess.check_output(cmd, shell=True).decode('ascii')[:-1]
 
     @classmethod
     def get_network_interface_state(cls, interface):
+        '''
+        Extracts the current operation state of a network interface from sys-fs
+        '''
         return subprocess.check_output('cat /sys/class/net/%s/operstate' % interface, shell=True).decode('ascii')[:-1]
+
+
+class SSD1306Driver:
+    '''
+    Custom SSD1306 display driver that works with Donkey's Python package selection
+    and implements time-constant i2c access based on queuing.
+    Inspired by https://github.com/BLavery/lib_oled96/blob/master/lib_oled96.py
+    '''
+
+    def __init__(self, address, i2c, process_limit, display_width, display_height):
+        if i2c is None:
+            import Adafruit_GPIO.I2C as I2C
+            i2c = I2C
+        self.device = i2c.get_i2c_device(address)
+        self.cmd_mode = 0x00
+        self.data_mode = 0x40
+        self.addr = address
+        self.process_limit = process_limit
+        self.width = display_width
+        self.height = display_height
+        self.pages = int(self.height / 8)
+        self.image = Image.new('1', (self.width, self.height))
+        # this is a "draw" object for preparing display contents
+        self.canvas = ImageDraw.Draw(self.image)
+        self.data_queue = Queue()
+        self.cmd_queue = Queue()
+        # iInitialize the display using the magic sequence
+        self._command(
+            self.Cmds.DISPLAYOFF,
+            self.Cmds.SETDISPLAYCLOCKDIV, 0x80,
+            self.Cmds.SETMULTIPLEX,       0x3F,
+            self.Cmds.SETDISPLAYOFFSET,   0x00,
+            self.Cmds.SETSTARTLINE,
+            self.Cmds.CHARGEPUMP,         0x14,
+            self.Cmds.MEMORYMODE,         0x00,
+            self.Cmds.SEGREMAP,
+            self.Cmds.COMSCANDEC,
+            self.Cmds.SETCOMPINS,         0x12,
+            self.Cmds.SETCONTRAST,        0xCF,
+            self.Cmds.SETPRECHARGE,       0xF1,
+            self.Cmds.SETVCOMDETECT,      0x40,
+            self.Cmds.DISPLAYALLON_RESUME,
+            self.Cmds.NORMALDISPLAY,
+            self.Cmds.DISPLAYON)
+        self.process_queues_at_once()
+        self.cls()
+        if display_height > 32:
+            self._show_donkey_logo()
+        self.process_queues_at_once()
+        print("OLED display initialized.")
+
+    def get_canvas(self):
+        return self.canvas
+
+    def process_queues(self):
```

Comment:
Out of curiosity, are there any actionable differences in performance when multiple i2c devices are involved ? Usually actuators are on a different bus, and every bus is modeled as a file handle in the kernel. 

Do you have any real benchmarks that measure the difference ? I have been using the `OLEDPart` and I have not detected any issues thus far.

Reply:
Hello @tikurahul,
thank you for your question - they made me think a lot about the design. First, I fully agree to you: the kernel models each i2c access as a own device. Thus, OLED drivers are totally fine in running as a own thread, when other thread use i2c at the same time. Hence, I moved the OLED part into an own thread to speed up the main drive loop with the latest commit.

Following your suggestions, I compared the performance of the existing part and the new OLED part on my Raspberry-Pi 3+ using the 128x32 mode from start up to idle with no loaded h5-model.

Original OLED as baseline
```
+--------------------+--------+-------+--------+--------+--------+--------+--------+
|        part        |  max   |  min  |  avg   |  50%   |  90%   |  99%   | 99.9%  |
+--------------------+--------+-------+--------+--------+--------+--------+--------+
|      OLEDPart      | 164.69 | 88.45 | 145.21 | 152.76 | 156.82 | 162.16 | 164.44 |
+--------------------+--------+-------+--------+--------+--------+--------+--------+
```

Implementation in this PR:
```
+--------------------+-------+------+-------+------+-------+-------+-------+
|        part        |  max  | min  |  avg  | 50%  |  90%  |  99%  | 99.9% |
+--------------------+-------+------+-------+------+-------+-------+-------+
|      OLEDPart      | 64.53 | 0.09 | 16.17 | 0.10 | 51.88 | 53.69 | 63.02 |
+--------------------+-------+------+-------+------+-------+-------+-------+
```
On average this the implementation is ~9 times faster. but also the 99% percentile looks promising. This is no suprise, since the low-level stuff is done directly in the part.

Interesting is what happens when the i2c bus gets under heavy load, e.g. when rendering graphics in 128x64 mode in this OLED implementation. If the queue size is set unlimited, the actuators react slow, since the Raspberry-PI hardware cannot dispatch all i2c calls fast enough due to limited system resources. If the process limit size is set to 32 i2c commands per thread call, actuators react smoothly while the display graphic is rendered.

Conclusion:
there is no issue with the existing OLED part. It is doing it's job fast enough and running it in an own thread is totally fine. Thus, I rename the PR. If graphic rendering should be supported on weaker platforms (e.g. Raspi 3+), i2c command queuing and optimized display primitives are needed to have smooth actuator movement. I leave it now open to you and the the other maintainers, if this part is worth the integration. It is quite sophisticated and may not follow the intension of Donkey car that thrives for an simple and straight forward design. Thank your for your time and effort you were spending on this. :-)